frontend/lib/resolution/scope-types.cpp

```diff
@@ -42,6 +42,86 @@ std::string IdAndFlags::flagsToString(Flags flags) {
   return ret;
 }
 
+using Flags = IdAndFlags::Flags;
+using FlagSet = IdAndFlags::FlagSet;
+
+FlagSet FlagSet::singleton(Flags flags) {
+  FlagSet toReturn;
+  toReturn.insert(flags);
+  return toReturn;
+}
+
+FlagSet FlagSet::empty() {
+  return FlagSet();
+}
+
+void FlagSet::insert(Flags excludeFlags) {
+  // booleans, like all lattices, follow the absorption law:
+  //
+  //     a /\ (a \/ b) = a   and   a \/ (a /\ b) = a
+  //
+  // Since Flags elements represent conjunction, if a & b = a,
+  // we know that b has all of flags in a, and maybe more. Thus,
+  // logically, b = a /\ b', where b' is the "more". But then, by the
+  // absorption law,
+  //
+  //     a \/ b = a \/ (a /\ b') = a
+  //
+  // In other words, if a & b = a, then only a needs to be in the
+  // FlagSet (which is a disjunction of flags). First try finding such a pair,
+  // so that we can keep the size of the set small.
+  for (auto& otherFlags : flagVec) {
+    if ((otherFlags & excludeFlags) == otherFlags) {
+      // excludeFlags is subsumed, we're done.
+      return;
+    }
+    if ((excludeFlags & otherFlags) == excludeFlags) {
+      // Existing entry is subsumed.
+      otherFlags = excludeFlags;
+      return;
+    }
+  }
```

Comment:
There is another situation in which we can simplify. I don't know if it comes up in practice in our scope resolver but it might be worth implementing anyway.

If we have a FlagSet representing A ⋃ B ⋃ C ⋃ ..., and we seek to add NOT_B, there is no point in creating B ⋃ NOT_B ⋃ ..., since that just matches everything. Instead, we can replace the current FlagSet with an empty FlagSet (which matches everything, right?). This rule only applies to singleton conjunctions, though. (I.e. if it were X ⋂ Y and we insert NOT_X, this rule cannot simplify it, so we should end up with (X ⋂ Y) ⋃ NOT_X).

Reply:
I think I'm going to note this as a comment for further optimization. Detecting singleton cases seems fairly expensive (needs a loop or a large-ish lookup table), and it's not clear if the optimization is worth it.